### PR TITLE
Write SAN parser

### LINKED ISSUE #4: Implement Position.parseSAN
We need the ability to parse SAN for a `Position`.

Implement a method in `Position` class, whose signature is `Move? parseSAN(String san)`. The move returned should be a legal move of the position, if any.

It should support crazyhouse and drop moves too.

Example implementation on: https://github.com/niklasf/chessops/blob/f287455e1cc83925c9f759f3c8829e2ca89d243f/src/san.ts#L82

### PR DESCRIPTION
**Pull request by dignissimus**
Resolves #4

### CODE REVIEW

File: lib/src/position.dart [hunk 1]
```diff
@@ -213,6 +213,208 @@ abstract class Position<T extends Position<T>> {
...
+
+    final firstAnnotationIndex = san.indexOf(RegExp('[!?#+]'));
+    if (firstAnnotationIndex != -1) {
+      san = san.substring(0, firstAnnotationIndex);
+    }
+
+    if (san == "O-O") {
+      Move? move;
+      if (turn == Side.white) {
+        move = NormalMove(from: parseSquare("e1")!, to: parseSquare("g1")!);
```
Comment (veloce) [inline lib/src/position.dart:231]: that's preferable to use the square `int` value directly here.

File: lib/src/position.dart [hunk 2]
```diff
@@ -213,6 +213,208 @@ abstract class Position<T extends Position<T>> {
...
+      if (turn == Side.black) {
+        move = NormalMove(from: parseSquare("e8")!, to: parseSquare("c8")!);
+      }
+      if (!isLegal(move!)) {
+        return null;
+      }
+      return move;
+    }
+
+    final isPromotion = san.contains("=");
```
Comment (veloce) [inline lib/src/position.dart:255]: I'd rather have single quotes string for consistency in the project. I just added a lint rule on main to help with this.

File: lib/src/position.dart [hunk 3]
```diff
@@ -213,6 +213,208 @@ abstract class Position<T extends Position<T>> {
...
+        // Ambiguous notation
+        return null;
+      }
+    }
+
+    if (move == null) {
+      return null;
+    }
+
+    return move;
```
Comment (veloce) [inline lib/src/position.dart:415]: I don't think it supports `DropMove` and Crazyhouse variant right now. It should support them.

File: test/position_test.dart [hunk 4]
```diff
@@ -129,6 +129,90 @@ void main() {
         expect(sans.item2,
             equals(['e4', 'e5', 'Bc4', 'Nc6', 'Qf3', 'd6', 'Qxf7#']));
       });
```
Comment (veloce) [inline test/position_test.dart:131]: On top of the tests you already wrote, I'd like to have all the tests of the chessops package as well:

https://github.com/niklasf/chessops/blob/f287455e1cc83925c9f759f3c8829e2ca89d243f/src/san.test.ts#L42-L202

File: lib/src/position.dart [hunk 5]
```diff
@@ -213,6 +213,254 @@ abstract class Position<T extends Position<T>> {
...
+        // Invalid SAN
+        if (san[1] != 'x') {
+          return null;
+        }
+
+        // Remove the source file character and the capture marker
+        san = san.substring(2);
+      }
+
+      if (isPromotion) {
```
Comment (veloce) [inline lib/src/position.dart:331]: Please add some tests that validate promotion parsing.

Also here I think it miss Antichess support. See: https://github.com/niklasf/chessops/blob/f287455e1cc83925c9f759f3c8829e2ca89d243f/src/san.ts#L126

Please also add a test that validates Antichess promotion.

**Review by veloce — COMMENTED** [review]
Looks good but we're still missing some cases.

File: lib/src/position.dart [hunk 6]
```diff
@@ -213,6 +213,254 @@ abstract class Position<T extends Position<T>> {
...
+        return null;
+      }
+      final move = DropMove(to: destination, role: role);
+      if (!isLegal(move)) {
+        return null;
+      }
+      return move;
+    }
+
+    if (san == 'O-O') {
```
[the remark this answers is not in the document]
Comment (dignissimus) [reply lib/src/position.dart:259]: These are already supported due to lines 223 to 226. I will add tests today

File: lib/src/position.dart [hunk 7]
```diff
@@ -44,6 +44,9 @@ abstract class Position<T extends Position<T>> {
   /// Current move number.
   final int fullmoves;
 
+  /// Whether this is a position from an Antichess game
+  bool get isAntichess => false;
```
Comment (veloce) [inline lib/src/position.dart:48]: you don't need this; instead use:
```dart
    if (this is Antichess) ...
```

directly

**Pull request merged**